### ISSUE TITLE
LDAP Thread Leak

### ISSUE THREAD

**Issue by llsand** [floragunncom/search-guard#704]
* Search Guard and Elasticsearch version: 6.6.1-24.1
* JVM version and operating system version: elasticsearch-docker-image: 
docker.elastic.co/elasticsearch/elasticsearch:6.6.1
* Number of nodes in your cluster: 4

We are facing problems with Threads leaked by LDAP-Connections.
After 3-4 hours every cluster-nodes errors out with:

[2019-06-24T14:11:21,463][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [elasticsearch1] fatal error in thread [elasticsearch[elasticsearch1][write][T#2]], exiting
java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached
        at java.lang.Thread.start0(Native Method) ~[?:?]
        at java.lang.Thread.start(Thread.java:803) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:937) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.processWorkerExit(ThreadPoolExecutor.java:1005) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]

We have set linux-limits according to the documentation.
Moitoring the linux-process we see an ongoing increae of threads (with  ps huH p 18039 | wc -l)

A kill -3 on the process reveals many thredas like this one:
"Thread-185" #294 daemon prio=5 os_prio=0 cpu=0.13ms elapsed=244.03s tid=0x00002b16fc319000 nid=0x19e runnable  [0x00002b17946f0000]
   java.lang.Thread.State: RUNNABLE
        at java.net.SocketInputStream.socketRead0(java.base@11.0.1/Native Method)
        at java.net.SocketInputStream.socketRead(java.base@11.0.1/SocketInputStream.java:115)
        at java.net.SocketInputStream.read(java.base@11.0.1/SocketInputStream.java:168)
        at java.net.SocketInputStream.read(java.base@11.0.1/SocketInputStream.java:140)
        at java.io.BufferedInputStream.fill(java.base@11.0.1/BufferedInputStream.java:252)
        at java.io.BufferedInputStream.read1(java.base@11.0.1/BufferedInputStream.java:292)
        at java.io.BufferedInputStream.read(java.base@11.0.1/BufferedInputStream.java:351)
        - locked <0x0000000627e45478> (a java.io.BufferedInputStream)
        at com.sun.jndi.ldap.Connection.run(java.naming@11.0.1/Connection.java:832)
        at java.lang.Thread.run(java.base@11.0.1/Thread.java:834)

**Comment by floragunncom** [comment]
Can you please post your sg_config.yml and elasticsearch.yml?

**Comment by llsand** [comment]
config files:

[configfiles.zip](https://github.com/floragunncom/search-guard/files/3325031/configfiles.zip)

**Comment by floragunncom** [comment]
Are you running the cluster on Java 8 or 11?

It seems you hit an issue with was fixed with SG 24.2 so we recommend to either upgrade to SG 24.3 (which is the last fix version for 24) or to SG 25.1 (which is the most current release but only available for ES 6.6.2 or higher).

**Comment by llsand** [comment]
The java version is the included with the docker image.
I think it is 11.

**Comment by floragunncom** [comment]
Thx. Please upgrade as explained above and report back if this fix your issue

**Comment by llsand** [comment]
Looks like the probelm disappears with 6.6.1-24.3.
Thank you very much!

**Comment by floragunncom** [comment]
There exists indeed a connection leak. This is due to a JVM bug for Java >= 9, see here https://github.com/vt-middleware/ldaptive/issues/148 for more details.

The reason why it works with 6.6.1-24.3 is because we fixed an issue with user caching in 24.2 and when caching is used (and is working properly) the leak does not become effective under normal circumstances. We fixed it nevertheless of course.

Expect a fix in the next version.
